File: src/utils/adjustForShrinkandOffset.ts
import { atan2, cos, sin, sqrt } from "mathjs";

import type { RawPoint } from "./blendMerge";

export function adjustForShrinkAndOffset(
	pointLayers: RawPoint[][],
	center: RawPoint,
): RawPoint[][] {
	const store = window.provelPrintStore;
	const activeProfile = window.materialProfiles[store.activeMaterialProfile];
	const nozzleSize = store.nozzleSize;
	const shrinkAllowance = activeProfile.shrinkFactor; // assumed to be in percentage

	if (nozzleSize === 0 || shrinkAllowance === 0) {
		return pointLayers;
	}

	return pointLayers.map((layer) =>
		layer.map((pt) => {
			// Shift the point relative to the center
			const dx = pt.x - center.x;
			const dz = pt.z - center.z;

			// Get polar coordinates from the shifted point (projected on X-Z plane)
			const r = sqrt(dx * dx + dz * dz) as number;
			const theta = atan2(dz, dx);

			// Adjust radius: add nozzle size and apply shrink percentage adjustment
			let adjustedR = r + nozzleSize;
			adjustedR = adjustedR * (1 - shrinkAllowance / 100);

			// Create new adjusted point and shift it back to the original coordinate system
			return {
				x: center.x + adjustedR * cos(theta),
				y: pt.y,
				z: center.z + adjustedR * sin(theta),
			};
		}),
	);
}

File: src-tauri/src/adjust_for_shrink_and_offset.rs
use std::f64::consts::PI;
use std::sync::Arc;

#[derive(Clone, Copy, Debug)]
struct RawPoint {
    x: f64,
    y: f64,
    z: f64,
}

async fn get_nozzle_size() -> f64 {
    // Simulate database call
    0.4 // Example nozzle size
}

async fn get_active_material_profile() -> MaterialProfile {
    // Simulate fetching material profile
    MaterialProfile { shrink_factor: 2.0 } // Example shrink factor in percentage
}

#[derive(Debug)]
struct MaterialProfile {
    shrink_factor: f64,
}

pub async fn adjust_for_shrink_and_offset(
    points: Vec<Vec<RawPoint>>,
    center: RawPoint,
) -> Vec<Vec<RawPoint>> {
    let active_material_profile = get_active_material_profile().await;
    let nozzle_size = get_nozzle_size().await;
    let shrink_allowance = active_material_profile.shrink_factor;

    if nozzle_size == 0.0 || shrink_allowance == 0.0 {
        return points;
    }

    let mut adjusted_points = Vec::new();

    for layer in points.iter() {
        let mut adjusted_layer = Vec::new();
        for &pt in layer.iter() {
            let dx = pt.x - center.x;
            let dz = pt.z - center.z;

            let r = (dx * dx + dz * dz).sqrt();
            let theta = dz.atan2(dx);

            let mut adjusted_r = r + nozzle_size / 2.0;
            adjusted_r *= 1.0 - shrink_allowance / 100.0;

            adjusted_layer.push(RawPoint {
                x: center.x + adjusted_r * theta.cos(),
                y: pt.y,
                z: center.z + adjusted_r * theta.sin(),
            });
        }
        adjusted_points.push(adjusted_layer);
    }

    adjusted_points
}
